Retrospettiva (#114)
anitvam/pps-bunny: doc/report.md

e `it.unibo.pps.bunny.model.world` che contiene la modellazione dell'ambiente di simulazione e della riproduzione.

È fornito di seguito uno screenshot che evidenzia le informazioni sopra riportate, calcolate con l'ausilio del
plugin `sbt-scoverage`:

![](images/scoverage_report_2.png)

L'uso della CI su GitHub unito a tale livello di coverage garantisce che sul branch `develop` non sia mai stato inserito
del codice non correttamente funzionante, infatti se i test non passano, la pull request effettuata dallo sviluppatore
non viene mergiata.

Per quanto riguarda invece il package `it.unibo.pps.bunny.view` non sono stati implementati test di `ScalaTest` per
verificarne il funzionamento, in quanto si è deciso di effettuare prevalentemente del beta-testing, cioè di eseguire
direttamente l'applicativo per verificare ad esempio il movimento dei coniglietti e la gestione dei pannelli per l'
inserimento delle mutazioni e dei fattori, oltre che per avere una controprova visiva del corretto funzionamento delle
classi di model e dell'engine.

## Retrospettiva

In questa sezione saranno descritti gli Sprint settimanali con le rispettive relazioni e i commenti finali del team.

> Inoltre nel file Backlog della cartella `/process` del repository GitHub sono riportati i riassunti degli Sprint.

Reviewer: @alessiarocco98 ti avevo lasciato anche il commento sopra, potresti aggiungere il file che citi? senò ci dimentichiamo e non è il massimo avere scritto un riferimento non vero

> Inoltre nel file Backlog della cartella `/process` del repository GitHub sono riportati i riassunti degli Sprint.

Author: certo assolutamente! Preferiamo riportare tutti gli Sprint Planning in un unico file o un file per ogni Sprint? @anitvam @aslucc @martaSpadoni

> Inoltre nel file Backlog della cartella `/process` del repository GitHub sono riportati i riassunti degli Sprint.

Reviewer 2: facciamo noi al volo tranquilla

### Sprint

> Sono stati realizzati un totale di 5 Sprint, alla conclusione di ognuno il team si è riunito per rilasciare un prototipo
> funzionante, ad eccezione del primo Sprint e per discutere del lavoro svolto, congiuntamente si è anche provveduto ad eseguire lo Sprint Planning dello Sprint successivo.

Reviewer: ```suggestion
Sono stati realizzati un totale di 5 Sprint, alla conclusione di ognuno il team si è riunito per rilasciare un prototipo
funzionante (ad eccezione del primo Sprint) e per discutere del lavoro svolto, congiuntamente si è anche provveduto ad eseguire lo Sprint Planning dello Sprint successivo.
```

#### Sprint 1

> Durante il primo sprint, il team si è concentrato sulla fase di analisi del dominio, parte fondamentale per il futuro
> sviluppo del progetto. Ogni membro del gruppo, in autonomia, ha prodotto un diagramma UML ad alto livello del model, tali diagrammi sono poi stati discussi e uniti per generare il diagramma su cui successivamente si è basata l'intera implementazione del Model. Si è inoltre iniziato a compilare il Product Backlog, a definire una prima suddivisione dei task e a preparare il repository del progetto, configurando SBT e la CI

Reviewer: ```suggestion
Durante il primo sprint, il team si è concentrato sulla fase di analisi del dominio, parte fondamentale per il futuro
sviluppo del progetto. Ogni membro del gruppo, in autonomia, ha prodotto un diagramma UML ad alto livello del model, tali diagrammi sono poi stati discussi e uniti per generare lo schema su cui successivamente si è basata l'intera implementazione del Model. Si è inoltre iniziato a compilare il Product Backlog, a definire una prima suddivisione dei task e a preparare il repository del progetto, configurando SBT e la CI.
```

#### Sprint 2
L'obiettivo del secondo Sprint era quello di implementare un primo prototipo funzionante del sistema. Quest'ultimo, basandosi sull'architettura MVC, doveva presentare: la parte di model relativa ai Bunny e al genoma, la componente dell'engine per definire il loop di avanzamento delle generazioni e, lato View, la struttura principale dell'interfaccia grafica e la possibilità di visualizzare i coniglietti nel pannello principale.

Il prototipo sottomesso al termine dello sprint presenta dunque una versione dell'interfaccia grafica minimale all'interno della quale è possibile visionare la riproduzione del coniglietto senza l'introduzione di mutazioni.

L'obiettivo è stato portato a termine senza particolari difficoltà, grazie anche alla buona fase di analisi condotta precedentemente.

#### Sprint 3

Il terzo Sprint ha avuto come obiettivo l'espansione del prototipo precedentemente prodotto aggiungendo la gestione delle mutazioni sia lato Model che View e una prima versione di ogni grafico del sistema.

> Durante questo Sprint Planning il team era indeciso se implementare prima la parte dei fattori o quella delle mutazioni, a seguito di un'attenta valutazione si però deciso di procedere come descritto in quanto avere già a disposizione la mutazioni nei coniglietti sarebbe stato utile successivamente nella definizione dei Fattori, in particolar modo per la parte di testing di quest'ultimi. A posteriori il team conferma la correttezza della scelta presa.

Reviewer: ```suggestion
Durante questo Sprint Planning il team era indeciso se implementare prima la parte dei fattori o quella delle mutazioni, a seguito di un'attenta valutazione si è scelto di procedere come descritto in quanto avere a disposizione la mutazioni nei coniglietti sarebbe stato utile alla successiva definizione dei fattori, in particolar modo per il testing di quest'ultimi. A posteriori il team si ritiene soddisfatto della scelta compiuta.
```

### Commenti finali

Per concludere, si riporta di seguito un commento che riguarda il sistema finale creato e l'esperienza
maturata da tutto il team durante il processo di sviluppo.
Il lavoro è stato svolto rispettando in buona parte gli standard principali della metodologia agile e di Scrum, in particolar modo attraverso il rilascio di un prototipo funzionante alla fine di ogni Sprint fissato. Le scadenze sono state rispettate nella maggior parte dei casi, con'unica eccezione alla conclusione del quarto Sprtint, avvenuta con un giorno di ritardo. Nel complesso non sono state riscontrate problematiche particolari relative alla pianificazione e la metodologia di sviluppo.
A posteriori, il team si ritiene sufficientemente soddisfatto dal lavoro svolto in quanto è stata sviluppata la totalità delle funzionalità obbligatorie e parte di quelle facoltative nei limiti di tempo predefiniti.

#### Sviluppi futuri

> Per quanto riguarda gli sviluppi futuri il gruppo propone i seguenti aspetti per l'aggiunta di ulteriori funzionalità
> alla simulazione realizzata:
> * Aggiunta di nuove mutazioni;
> * Aggiunta di nuovi geni;
> * Aggiunta di nuove tipologie di fattori disturbanti;
> * Aggiunta di nuovi grafici;
> * Aggiunta di nuovi ambienti oltre al clima caldo e freddo;
> * Estensione e/o modifica delle fasi durante una generazione;
> * Estensione del simulatore ad una tipologia di animale diversa;

Reviewer: ```suggestion
```
Qui si può togliere che l'ho integrata nel nuovo branch

riportate le attività svolte durante il terzo meeting:

* Chiusura di tutti i task Sprint 2
* Merge di tutte le Pull Request relative allo Sprint 2
* Rilascio Release v0.1.0
* Revisione difficoltà dei task
* Product Backlog per Sprint 3 e 4
* Sprint Planning 3 con suddivisione dei task
* Definizione dei pannelli necessari nella GUI, per le mutazioni, per i grafici e l'albero genealogico
* Definizione e introduzione dell'aspetto delle mutazioni
* Definizione iniziale del report

#### Sprint 4

Durante il quarto Sprint il team si riunito prima di tutto per il meeting settimanale nel quale si è discusso sia la
fine dello Sprint precedente che la definizione dello Sprint Planning del quarto. Di seguito sono riportate le attività
svolte durante il quarto meeting:

* Revisione difficoltà dei task
* Chiusura di tutti i task Sprint 3
* Merge di tutte le Pull Request relative allo Sprint 3
* Rilascio Release v.0.2.0
* Aggiornamento del Product Backlog per Sprint 4
* Sprint Planning 4
* Definizione della realizzazione dei fattori disturbanti
* Definizione del restart e cambiamento di velocità della simulazione

> * Definizione della realizzazione della grafica dell'orologio

Reviewer 3: Come per i due Sprint precedenti anche per il quarto, il team si è posto l'obiettivo di realizzare un prodotto concreto e in questo caso, il prototipo è molto vicino alla versione finale del sistema. Nello specifico, si sono introdotti i fattori disturbanti, sia lato Model che View e si è aggiunta la possibilità di modificare la velocità del loop e di eseguire il restart al termine di ogni simulazione. Si è inoltre voluto inserire un nuovo componente nell'interfaccia grafica: un orologio, questo permette all'utente di comprendere in che momento della generazione si trova durante la simulazione.

> Durante il quarto Sprint il team si riunito prima di tutto per il meeting settimanale nel quale si è discusso sia la
> fine dello Sprint precedente che la definizione dello Sprint Planning del quarto. Di seguito sono riportate le attività
> svolte durante il quarto meeting:
> 
> * Revisione difficoltà dei task
> * Chiusura di tutti i task Sprint 3
> * Merge di tutte le Pull Request relative allo Sprint 3
> * Rilascio Release v.0.2.0
> * Aggiornamento del Product Backlog per Sprint 4
> * Sprint Planning 4
> * Definizione della realizzazione dei fattori disturbanti
> * Definizione del restart e cambiamento di velocità della simulazione
> * Definizione della realizzazione della grafica dell'orologio

Reviewer 3: ```suggestion
Come per i due Sprint precedenti anche per il quarto, il team si è posto l'obiettivo di realizzare un prodotto concreto e in questo caso, il prototipo è molto vicino alla versione finale del sistema. Nello specifico, si sono introdotti i fattori disturbanti, sia lato Model che View e si è aggiunta la possibilità di modificare la velocità del loop e di eseguire il restart al termine di ogni simulazione. Si è inoltre voluto inserire un nuovo componente nell'interfaccia grafica: un orologio, questo permette all'utente di comprendere in che momento della generazione si trova durante la simulazione.
```

Reviewer 2: Ho modificato un pochetto il commento

#### Sprint 5

Durante l'ultimo Sprint il team si è riunito per la riunione iniziale nella quale si è discussa sia la fine del
precedente Sprint che lo Sprint Planning del quinto. Non sono ha portato allo sviluppo di nuove feature ma si è
focalizzato nel miglioramento di tutto il lavoro svolto con la prospettiva di rendere codice il più leggibile e
riusabile possibile. Di seguito sono riportate le attività svolte durante l'ultimo meeting:

* Revisione difficoltà dei task
* Chiusura dei task di Sprint 4 e spostamento del task del clock nello sprint 5
* Merge di tutte le Pull Request relative allo Sprint 4 tranne quella relativa al Clock e alla correzione del problema
  di gui relativo al task della visualizzazione dei lupi.
* Rilascio Release v.0.3.0
* Aggiornamento del Product Backlog per Sprint 5
* Sprint Planning 5
* Revisione finale del codice con refactoring e scaladoc

> * Stesura finale del report

Reviewer 2: ho modificato un pochetto il commento

* Aggiornamento del Product Backlog per Sprint 4
* Sprint Planning 4
* Definizione della realizzazione dei fattori disturbanti
* Definizione del restart e cambiamento di velocità della simulazione
* Definizione della realizzazione della grafica dell'orologio

#### Sprint 5

Durante l'ultimo Sprint il team si è riunito per la riunione iniziale nella quale si è discussa sia la fine del
precedente Sprint che lo Sprint Planning del quinto. Non sono ha portato allo sviluppo di nuove feature ma si è
focalizzato nel miglioramento di tutto il lavoro svolto con la prospettiva di rendere codice il più leggibile e
riusabile possibile. Di seguito sono riportate le attività svolte durante l'ultimo meeting:

* Revisione difficoltà dei task
* Chiusura dei task di Sprint 4 e spostamento del task del clock nello sprint 5
* Merge di tutte le Pull Request relative allo Sprint 4 tranne quella relativa al Clock e alla correzione del problema
  di gui relativo al task della visualizzazione dei lupi.
* Rilascio Release v.0.3.0
* Aggiornamento del Product Backlog per Sprint 5
* Sprint Planning 5
* Revisione finale del codice con refactoring e scaladoc
* Stesura finale del report

### Commenti finali

> Per concludere, si riportano di seguito i commenti finali che tengono conto del sistema finale creato e dall'esperienza
> maturata da tutto il team durante il processo di sviluppo.
> 
> Il lavoro è stato compiuto rispettando fedelmente gli standard più salienti della metodologia agile e di Scrum, in modo
> particolare nel rilascio di un prototipo funzionante alla fine di ogni Sprint fissato. Tutto sommato il team ha
> rispettato le scadenze, tranne che per il rilascio del prodotto a seguito del penultimo Sprint che è avvenuto con un
> giorno di ritardo. Nel complesso all'interno del team non sono state riscontrate grosse problematiche relative alla
> pianificazione e metodologia di sviluppo.
> 
> Considerando a posteriori il lavoro svolto il team si ritiene pienamente soddisfatto in quanto è stata portata a termine
> la totalità delle funzionalità dichiarate nei limiti di tempo predefiniti, senza slittamenti temporali.

Reviewer: ```suggestion
Per concludere, si riporta di seguito un commento che riguarda il sistema finale creato e l'esperienza
maturata da tutto il team durante il processo di sviluppo.

Il lavoro è stato svolto rispettando in buona parta gli standard principali della metodologia agile e di Scrum, in particolar modo attraverso il rilascio di un prototipo funzionante alla fine di ogni Sprint fissato. Le scadenze sono state rispettate nella maggior parte dei casi, con'unica eccezione alla conclusione del quarto Sprtint, avvenuta con un giorno di ritardo. Nel complesso non sono state riscontrate problematiche particolari relative alla pianificazione e la metodologia di sviluppo.

A posteriori, il team si ritiene suffiecientemente soddisfatto dal lavoro svolto in quanto è stata sviluppata la totalità delle funzionalità obbligatorie e parte di quelle facoltative nei limiti di tempo predefiniti.
```

> Per concludere, si riportano di seguito i commenti finali che tengono conto del sistema finale creato e dall'esperienza
> maturata da tutto il team durante il processo di sviluppo.
> 
> Il lavoro è stato compiuto rispettando fedelmente gli standard più salienti della metodologia agile e di Scrum, in modo
> particolare nel rilascio di un prototipo funzionante alla fine di ogni Sprint fissato. Tutto sommato il team ha
> rispettato le scadenze, tranne che per il rilascio del prodotto a seguito del penultimo Sprint che è avvenuto con un
> giorno di ritardo. Nel complesso all'interno del team non sono state riscontrate grosse problematiche relative alla
> pianificazione e metodologia di sviluppo.
> 
> Considerando a posteriori il lavoro svolto il team si ritiene pienamente soddisfatto in quanto è stata portata a termine
> la totalità delle funzionalità dichiarate nei limiti di tempo predefiniti, senza slittamenti temporali.

Reviewer: ```suggestion
Per concludere, si riporta di seguito un commento che riguarda il sistema finale creato e l'esperienza
maturata da tutto il team durante il processo di sviluppo.
Il lavoro è stato svolto rispettando in buona parte gli standard principali della metodologia agile e di Scrum, in particolar modo attraverso il rilascio di un prototipo funzionante alla fine di ogni Sprint fissato. Le scadenze sono state rispettate nella maggior parte dei casi, con'unica eccezione alla conclusione del quarto Sprtint, avvenuta con un giorno di ritardo. Nel complesso non sono state riscontrate problematiche particolari relative alla pianificazione e la metodologia di sviluppo.
A posteriori, il team si ritiene sufficientemente soddisfatto dal lavoro svolto in quanto è stata sviluppata la totalità delle funzionalità obbligatorie e parte di quelle facoltative nei limiti di tempo predefiniti.
```

Per quanto riguarda invece il package `it.unibo.pps.bunny.view` non sono stati implementati test di `ScalaTest` per verificarne il funzionamento, in quanto si è deciso di effettuare prevalentemente del beta-testing, cioè di eseguire direttamente l'applicativo per verificare ad esempio il movimento dei coniglietti e la gestione dei pannelli per l'inserimento delle mutazioni e dei fattori, oltre che per avere una controprova visiva del corretto funzionamento delle classi di model e dell'engine.

## Retrospettiva

> 
> Nella sezione finale, la retrospettiva, verranno descritti gli Sprint settimanali con le rispettive relazioni e i
> commenti finali del team. Infine nel file Backlog nella cartella `/process` del repository GitHub sono riportati i
> riassunti degli Sprint.

Reviewer: ```suggestion

In questa sezione saranno descritti gli Sprint settimanali con le rispettive relazioni e i
commenti finali del team. Inoltre nel file Backlog della cartella `/process` del repository GitHub sono riportati i
riassunti degli Sprint.
```
Non ho capito la seconda frase, non vedo la cartella process

Author: dalle regole ho visto che dovremo poi mettere il Backlog nella cartella `/process` come file o di testo o un markdown. Per questo l'ho scritto

### Sprint

> Sono stati realizzati un totale di 5 Sprint della durata di due settimane ciascuno alla fine dei quali il team si è
> riunito per rilasciare un prototipo funzionante, ad eccezione del primo Sprint, per discutere del lavoro svolto,
> compilando il Product Backlog e il meeting report, e infine organizzando il lavoro futuro.

Reviewer: ```suggestion
Sono stati realizzati un totale di 5 Sprint, alla conclusione di ognuno il team si è riunito per rilasciare un prototipo funzionante, ad eccezione del primo Sprint. Inoltre si è discusso del lavoro svolto, compilando il Product Backlog e il meeting report, e infine si è organizzato il lavoro futuro.
```

In realtà è un po' una ripetizione del paragrafo iniziale

#### Sprint 1

Durante la prima fase il team si è concentrato sulla fase di analisi, molto corposa e fondamentale per il futuro

> sviluppo del progetto e la suddivisione dei task. Durante il primo meeting si sono prese le seguenti decisioni:

Reviewer: ```suggestion
sviluppo del progetto e la suddivisione dei task. Di seguito sono riportate le attività svolte durante il primo meeting:
```
Dato che non si tratta sempre di decisioni, metterei questa dicitura per tutti gli Sprint

* Aggiornamento del Product Backlog per Sprint 4
* Sprint Planning 4
* Definizione della realizzazione dei fattori disturbanti
* Definizione del restart e cambiamento di velocità della simulazione
* Definizione della realizzazione della grafica dell'orologio

#### Sprint 5

Durante l'ultimo Sprint il team si è riunito per la riunione iniziale nella quale si è discussa sia la fine del
precedente Sprint che lo Sprint Planning del quinto. Non sono ha portato allo sviluppo di nuove feature ma si è
focalizzato nel miglioramento di tutto il lavoro svolto con la prospettiva di rendere codice il più leggibile e
riusabile possibile. Durante il quinto meeting si sono prese le seguenti decisioni:

* Revisione difficoltà dei task
* Chiusura dei task di Sprint 4 e spostamento del task del clock nello sprint 5
* Merge di tutte le Pull Request relative allo Sprint 4 tranne quella relativa al Clock e alla correzione del problema
  di gui relativo al task della visualizzazione dei lupi.
* Rilascio Release v.0.3.0
* Aggiornamento del Product Backlog per Sprint 5
* Sprint Planning 5
* Revisione finale del codice con refactoring e scaladoc
* Stesura finale del report

// Commenti finali

> // Sviluppi futuri

Reviewer: Puoi provare ad abbozzare questa parte
